### ISSUE TITLE
Don't demote suggestions if the user continues to type text that matches.

### ISSUE THREAD

**Issue by najamelan** [codota/TabNine#231]
Thanks for this amazing tool.

I have one issue. Sometimes you are typing a bit fast and it takes a few letters to see that the suggestion is a good one. On top of that, often typing a few letters will let TabNine give better suggestions. 

A frustrating thing I have noted is that often, when I type the first letter, the suggestion will be exactly what I'm looking for, but then when I type letters 2 and 3, even though they correspond to what the suggestion was, it goes down in the list or disappears.

It seems counter-intuitive that a winning suggestion all of a sudden is no longer winning even though all characters typed match it. Easy to reproduce:

Type `fn`
![tabfn](https://user-images.githubusercontent.com/449589/76246724-2bc04b80-6236-11ea-824e-d7d7f060a153.png)

Type an additional space:
![tabspace](https://user-images.githubusercontent.com/449589/76246760-41357580-6236-11ea-877b-ace31e3dc1ac.png)

**Comment by avichay77** [comment]
Thanks for reporting this issue. We are reviewing this issue. If you have additional examples please share with us.

**Comment by najamelan** [comment]
I can, but it's really easy to reproduce. Like in the above screenshot it looks like the different suggestions in the second one come from a different provider, like RLS, but that's not the issue. it even happens with suggestions that are clearly of the same nature as in the first screenshot, the scoring just changes completely when extra characters are added.

**Comment by najamelan** [comment]
![tabnine-a](https://user-images.githubusercontent.com/449589/76529084-b210a380-6469-11ea-8537-399ce0b7c727.png)
![tabnine-as](https://user-images.githubusercontent.com/449589/76529086-b341d080-6469-11ea-9c7d-cbeb250bcdaa.png)

Here, when using backspace and retyping the `a`,  now `assert_matches` immediately is suggested first, so it changes over time.

When trying to reproduce, I also found that it doesn't happen 100% of the time. Sometimes the first suggestion stays first. It just looks like the algorithm is quite non-deterministic and a two character input doesn't build on the results from a one character input.

One situation where this is annoying as well is that TabNine suggests long strings and sometimes the first part is correct, but I don't want to hit tab because it is going to insert 50 chars of which I will have to backspace 40 to where I need to divert from the suggestion and get a better suggestion for the next part. So the solution is to keep typing until the part that diverges. Now I know this doesn't really change a thing, because the solution remains the same (keep typing) but it's kind of disturbing that the suggestion you are looking at and are typing in disappears. That's purely psychological though.

Maybe this warrants a separate issue, but in light of that last paragraph, I would say that if two suggestions have a common prefix, it's always better to show the shortest one with a higher score. When the user hits tab and inserts the shorter one, immediately show suggestions for the next part. In the next screenshot I have tabbed in the `assert_matches`. However, even though the suggestions window stays open, I will not get any meaningful suggestions until I start typing again (in this case, I'm pretty much guaranteed to type a `(`, and when I do, TabNine will provide new suggestions):
![tabnine-assert](https://user-images.githubusercontent.com/449589/76530974-657a9780-646c-11ea-8a00-170f467d737d.png)
